prometheus/vec_test.go

```diff
@@ -125,6 +125,87 @@ func testDeleteLabelValues(t *testing.T, vec *GaugeVec) {
 	}
 }
 
+func TestDeletePartialMatch(t *testing.T) {
+	vec := NewGaugeVec(
+		GaugeOpts{
+			Name: "test",
+			Help: "helpless",
+		},
+		[]string{"l1", "l2"},
+	)
+	testDeletePartialMatch(t, vec)
+}
+
+func testDeletePartialMatch(t *testing.T, vec *GaugeVec) {
+	// No metric value is set.
+	if got, want := vec.DeletePartialMatch(Labels{"l1": "v1", "l2": "v2"}), 0; got != want {
+		t.Errorf("got %v, want %v", got, want)
+	}
+
+	c1 := vec.MustCurryWith(Labels{"l1": "v1"})
+	c1.WithLabelValues("2").Inc()
+
+	// Try to delete nonexistent label lx with existent value v1.
+	if got, want := c1.DeletePartialMatch(Labels{"lx": "v1"}), 0; got != want {
+		t.Errorf("got %v, want %v", got, want)
+	}
+
+	// Delete with valid pair l1: v1.
+	if got, want := c1.DeletePartialMatch(Labels{"l1": "v1"}), 1; got != want {
+		t.Errorf("got %v, want %v", got, want)
+	}
+
+	// Try to delete with partially invalid labels.
+	vec.With(Labels{"l1": "v1", "l2": "v2"}).(Gauge).Set(42)
+	if got, want := vec.DeletePartialMatch(Labels{"l1": "v1", "l2": "xv2"}), 0; got != want {
+		t.Errorf("got %v, want %v", got, want)
+	}
+
+	// Try to delete with a single valid label which matches multiple metrics.
+	vec.With(Labels{"l1": "v1", "l2": "v2"}).(Gauge).Set(42)
+	vec.With(Labels{"l1": "v1", "l2": "vv22"}).(Gauge).Set(84)
+	c3 := vec.MustCurryWith(Labels{"l2": "l2C3CurriedValue"}) // Used below
+	vec.With(Labels{"l1": "v3", "l2": "v3"}).(Gauge).Set(168)
+	if got, want := vec.DeletePartialMatch(Labels{"l1": "v1"}), 2; got != want {
+		t.Errorf("got %v, want %v", got, want)
+	}
+
+	// Try to delete a value which shouldn't be in our base vector (only the curried one c3).
+	if got, want := vec.DeletePartialMatch(Labels{"l2": "l2C3CurriedValue"}), 0; got != want {
+		t.Errorf("got %v, want %v", got, want)
+	}
+
+	c2 := vec.MustCurryWith(Labels{"l2": "l2CurriedValue"})
+	c2.With(Labels{"l1": "11"}).Inc()
+
+	// Delete with valid curried pair l2: l2CurriedValue.
+	if got, want := c2.DeletePartialMatch(Labels{"l2": "l2CurriedValue"}), 1; got != want {
+		t.Errorf("got %v, want %v", got, want)
+	}
+
+	c3.With(Labels{"l1": "11"}).Inc()
+
+	// Try to delete with invalid curried pair l1: v1.
+	if got, want := c3.DeletePartialMatch(Labels{"l1": "v1"}), 0; got != want {
+		t.Errorf("got %v, want %v", got, want)
+	}
+	// Delete valid curried pair l2: l2C3CurriedValue.
+	if got, want := c3.DeletePartialMatch(Labels{"l2": "l2C3CurriedValue"}), 1; got != want {
+		t.Errorf("got %v, want %v", got, want)
+	}
+
+	// Try to delete with a label value from before currying.
+	if got, want := c2.DeletePartialMatch(Labels{"l2": "v3"}), 0; got != want {
```

Comment:
It's correct that this doesn't delete anything. But the reason is not that the label is "from before currying". The vector element `{"l1": "v3", "l2": "v3"}` still exists. It's just not accessible through the curried view (because that one hardcodes `l2` to `l2CurriedValue`, and therefore any attempt to specify the same or even a different value for `l2` should be considered invalid and not do anything).